### PR TITLE
Add Kotlin code generation implementation.

### PR DESCRIPTION
**Pull request by Leland-Takamine**
Enabled whenever the "kapt.kotlin.generated" or "motif.mode=kotlin" option is set.

### CODE REVIEW

**Comment by Leland-Takamine** [comment]
Base branch for this PR is `mode` right now but I'll rebase and update once #121 lands.

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                val className = JavaToKotlinClassMap.INSTANCE
+                        .mapJavaToKotlin(FqName(typeName.toString()))?.asSingleFqName()?.asString()
+                if (className == null) typeName
+                else com.squareup.kotlinpoet.ClassName.bestGuess(className)
+            }
+        }
+    }
+
+    /**
+     * Copied from [FunSpec.overriding] and modified to leverage [javaToKotlinType].
```
Comment (tyvsmith) [inline compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt]: If you copied this from kotlinpoet, we should update the license header to include square.

File: compiler/src/test/java/motif/compiler/TestHarness.java [hunk 2]
```diff
@@ -77,92 +95,146 @@ private static boolean isTestDir(File file) {
...
+            @Nullable Processor motifProcessor) {
+        KotlinCompilation compilation = new KotlinCompilation();
+        compilation.setAnnotationProcessors(getProcessors(motifProcessor));
+        compilation.setSources(sourceFiles.stream()
+                .map(SourceFile.Companion::fromPath)
+                .collect(Collectors.toList()));
+        compilation.setInheritClassPath(true);
+        compilation.setClasspaths(classpaths);
+        compilation.setKaptArgs(aptArgs);
+        compilation.setVerbose(false);
```
Comment (tyvsmith) [inline compiler/src/test/java/motif/compiler/TestHarness.java]: This should be a flag

File: compiler/src/test/java/motif/compiler/TestHarness.java [hunk 3]
```diff
@@ -77,92 +95,146 @@ private static boolean isTestDir(File file) {
...
+
+        return compilation.compile();
+    }
+
+    private List<File> getFiles(File dir) throws IOException {
+        return getFiles(dir, file -> (file.getName().endsWith(".java") || file.getName().endsWith(".kt")) && !file.getName().equals("ScopeImpl.java"));
```
Comment (tyvsmith) [inline compiler/src/test/java/motif/compiler/TestHarness.java]: kts?

File: compiler/src/test/java/motif/compiler/TestHarness.java [hunk 4]
```diff
@@ -29,45 +30,62 @@
...
+import java.nio.file.Path;
+import java.util.*;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
-import static com.google.testing.compile.CompilationSubject.assertThat;
+import static com.google.common.truth.Truth.assertWithMessage;
 
 @RunWith(Parameterized.class)
 public class TestHarness {
```
Comment (tyvsmith) [inline compiler/src/test/java/motif/compiler/TestHarness.java]: General question: Why is this class in Java, but most of the project is in Kotlin?

File: gradle/dependencies.gradle [hunk 5]
```diff
@@ -40,9 +40,10 @@ ext.deps = [
                 ]
         ],
         "kotlin": [
-                stdlib: "org.jetbrains.kotlin:kotlin-stdlib:${versions.kotlin}",
-                reflection: "org.jetbrains.kotlin:kotlin-reflect:${versions.kotlin}"
-
+                compiler: "org.jetbrains.kotlin:kotlin-compiler-embeddable:${versions.kotlin}",
```
Comment (tyvsmith) [inline gradle/dependencies.gradle]: Let's create a task and track about moving off of embeddable

**Comment by tyvsmith** [comment]
Are you planning on adding parsing of Kotlin files (and test kotlin files) to this branch as well, or is that separate?

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            returns(returnTypeName.kt)
+            addKdoc(javaDoc.spec())
+        }.build()
+    }
+
+    private fun Qualifier.spec(): AnnotationSpec {
+        return AnnotationSpec.get(annotation.mirror)
+    }
+
+    private fun DependencyMethodJavaDoc.spec(): CodeBlock {
```
Comment (mcassiano) [inline compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt]: Should we generate KDoc here instead of JavaDoc?

File: compiler/src/test/java/motif/compiler/TestHarness.java (hunk 2 above)
Comment (Leland-Takamine) [reply compiler/src/test/java/motif/compiler/TestHarness.java]: This is just part of the test harness and we will never use verbose mode here.

File: compiler/src/test/java/motif/compiler/TestHarness.java (hunk 3 above)
Comment (Leland-Takamine) [reply compiler/src/test/java/motif/compiler/TestHarness.java]: This is part of the test harness and we can assume all  kotlin files end with kt.

File: compiler/src/test/java/motif/compiler/TestHarness.java (hunk 4 above)
Comment (Leland-Takamine) [reply compiler/src/test/java/motif/compiler/TestHarness.java]: This is for legacy reasons. We can convert to kotlin in the future.

File: gradle/dependencies.gradle (hunk 5 above)
Comment (Leland-Takamine) [reply gradle/dependencies.gradle]: Sure - can you open an issue with a description of why moving off of embeddable is preferable?

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt (hunk 6 above)
Comment (Leland-Takamine) [reply compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt]: Ah right - I updated the link format but I suppose kdoc supports simple markdown lists?

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt (hunk 1 above)
Comment (Leland-Takamine) [reply compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt]: Got it will update

**Comment by Leland-Takamine** [comment]
> Are you planning on adding parsing of Kotlin files (and test kotlin files) to this branch as well, or is that separate?

That's a separate effort. There's a lot of edge cases to consider and we'll want to tackle that comprehensively.

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt (hunk 6 above)
Comment (mcassiano) [reply compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt]: Yup. I believe so.

Comment (Leland-Takamine) [reply compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt]: Updated

File: compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt (hunk 1 above)
Comment (Leland-Takamine) [reply compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinTypeWorkaround.kt]: Updated

File: build.gradle [hunk 7]
```diff
@@ -23,5 +23,6 @@ subprojects {
     repositories {
         jcenter()
         google()
+        maven { url 'https://jitpack.io' }
```
Comment (tyvsmith) [inline build.gradle]: Was this intentional? For the compileTestingKotlin dep?

Comment (Leland-Takamine) [reply build.gradle]: Yes this is required for the Kotlin testing library.

Comment (Leland-Takamine) [reply build.gradle]: I'll update with a comment here

Comment (Leland-Takamine) [reply build.gradle]: Updated
